### PR TITLE
fix FP in no-unused-array: for-in loop

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 1]
```diff
@@ -78,6 +78,9 @@ export class Rule extends tslint.Rules.TypedRule {
         // prettier-ignore
         const varDeclaration = firstAncestor(symbolAndDeclaration.declaration, [ts.SyntaxKind.VariableDeclaration]) as ts.VariableDeclaration;
         if (varDeclaration) {
+          if (is(varDeclaration.parent!.parent, ts.SyntaxKind.ForInStatement, ts.SyntaxKind.ForOfStatement)) {
```
Comment (SonarTech) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 2]
```diff
@@ -78,7 +78,10 @@ export class Rule extends tslint.Rules.TypedRule {
         // prettier-ignore
         const varDeclaration = firstAncestor(symbolAndDeclaration.declaration, [ts.SyntaxKind.VariableDeclaration]) as ts.VariableDeclaration;
         if (varDeclaration) {
-          return !varDeclaration.initializer || Rule.isArrayLiteral(varDeclaration.initializer);
+          if (is(varDeclaration.parent!.parent, ts.SyntaxKind.ForInStatement, ts.SyntaxKind.ForOfStatement)) {
```
Comment (SonarTech) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 3]
```diff
@@ -194,9 +202,14 @@ export class Rule extends tslint.Rules.TypedRule {
...
-      return (node as ts.NewExpression).expression.getText() === "Array";
+      const constructorName = (node as ts.NewExpression).expression.getText();
+      return Rule.isArrayMapOrSet(constructorName);
     }
 
     return false;
   }
+
+  private static isArrayMapOrSet(str: string): boolean {
+    return str === "Array" || str === "Set" || str === "Map"
```
Comment (inverno) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: Maybe consider supporting things like WeakMap or Uint8Array

**Pull request merged**